packages/react/src/components/Table/TableToolbar/TableToolbar.jsx

```diff
@@ -286,9 +291,18 @@ const TableToolbar = ({
               translateWithId={(...args) => tableTranslateWithId(i18n, ...args)}
               id={`${tableId}-toolbar-search`}
               onChange={(event, defaultValue) => {
-                // https://github.com/carbon-design-system/carbon/issues/6157
-                onApplySearch(event?.target?.value || defaultValue);
+                const value = event?.target?.value || defaultValue;
```

Comment:
I think that the "Clear search input" in TableToolbarSearch's updates the value to be an empty string and expects this onChange method to perform the actual searchApply to clear the filter. So for the new hasFastSearch:false the search will not be cleared when the "Clear search input" is clicked.